google/cloud/storage/internal/generate_message_boundary.h

```diff
@@ -33,13 +35,15 @@ namespace internal {
  * trivial to generate a string not found in @p message, just append some
  * characters to the message itself).
  *
- * The algorithm is to generate a short random string, and search for it in the
- * message, if the message has that string, append some more random characters
- * and keep searching.
- *
- * This function is a template because the string generator is typically a
- * lambda that captures state variables (such as the random number generator),
- * of the class that uses it.
+ * The algorithm is:
+ * - First, try using fixed strings consisting of 64 copies of any valid
+ *   boundary character. We can verify if one of those would work by searching
+ *   for the boundary character every 64 bytes in the message. If the character
+ *   never appears in that search, then there is no substring in the message
+ *   consisting of 64 copies of the character.
+ * - If that fails, generate a short random string, and search for it in the
+ *   message. If the message has that string, append some more random characters
+ *   and keep searching.
```

Comment:
Possible note for the future: Given that the message actually contained the length-n random string, it may be better to generate an entirely new length-n+k random string rather than appending k new random chars to the existing string.

Reply:
The advantage of adding `k` characters is that you don't need to search from the beginning.  If the first time you find `foo` is at position 1000, then you know  `foobar` is not found before position 1000 either. That means the algorithm is always making progress.